contracts/Dividends.sol

```
@@ -0,0 +1,110 @@
+
+    using SafeMath for uint256;
+
+    RestrictedToken restrictedToken;
+
+    event FundEvent (
+        address payer,
+        address token,
+        uint256 amount,
+        uint256 snapshotId
+    );
+
+    event ClaimEvent (
+        address payee,
+        address token,
+        uint256 amount,
+        uint256 snapshotId
+    );
+
+    mapping(uint256 => mapping(address => mapping(address => uint256))) fundersTokensGet; // snapshotID => funderAddress => token => getAmount;
+
+    mapping(uint256 => mapping(address => uint256)) tokensFunded; // snapshotID => token => totalAmount of token
+
+    constructor(address _restrictedToken) public {
+        restrictedToken = RestrictedToken(_restrictedToken);
+    }
+
+    function fundDividend(address _token, uint256 amount, uint256 snapshotId) public { // payable not used for tokens
+        IERC20 paymentToken = IERC20(_token);
+
+        require( _token != address(0), "WRONG TOKEN ADDRESS");
+
+        require(paymentToken.balanceOf(msg.sender) >= amount, "SENDER HAVE NOT TOKENS");
+
+        paymentToken.safeTransferFrom(msg.sender, address(this), amount);
+
+        if ( tokensFunded[snapshotId][_token] > 0 ) {
+            tokensFunded[snapshotId][_token].add(amount);
+        } else {
+            tokensFunded[snapshotId][_token] = amount;
+        }
+
+//        console.log("Token funded amount %s", amount);
+//        console.log("Token funded on deposit %s", tokensFunded[snapshotId][_token]);
+
+        emit FundEvent(msg.sender, _token, amount, snapshotId);
+    }
+
+    function fundedAt(address _token, uint256 snapshotId) public view returns (uint256) {
+        return uint256(tokensFunded[snapshotId][_token]);
+    }
+
+    function claimDividend(address _token, uint256 snapshotId) public {
+        // get balance of snapshotID
+        uint256 seqTokenBalance = restrictedToken.balanceOfAt(msg.sender, snapshotId);
+
+        require(seqTokenBalance > 0, "SECURITY TOKEN BALANCE IS NULL");
```

"SECURITY TOKEN BALANCE IS 0"

Is it better to raise an error here so it doesn't use any gas or is it better to just return without updating anything?

---

Personally, `require` seems to be a good solution

---

Personally, it seems to be a nice solution to return error